composer/templatetags/composer_tags.py

```diff
@@ -96,115 +77,92 @@ def tile(parser, token):
-            with context.push():
-                context['tile_target'] = tile.target
-                tag_file = "%s_tags" % tile.target.__module__.split(".")[0]
-                tag_string = '%s "%s"' % (
-                    tile.target.__class__.__name__.lower(),
-                    tile.target.slug)
-                try:
-                    t = template.Template("{%% load %s %%}{%% %s %%}" % (
-                        tag_file,
-                        tag_string))
-                    result = t.render(context)
-                except template.TemplateSyntaxError:
-                    t = template.Template("{%% load %s %%}{%% render_%s %%}" % (
-                        tag_file,
-                        tag_string))
-                    result = t.render(context)
-                return t.render(context)
-            return "The tile target could not be found"
-        return "The tile could not be rendered"
-
-
-@register.tag
-def tile_url(parser, token):
-    """Return the Url for a given view. Very similar to the {% url %} template
-    tag, but can accept a variable as first parameter."""
-    try:
-        tag_name, tile = token.split_contents()
-    except ValueError:
-        raise template.TemplateSyntaxError(
-            "tile_url tag requires argument tile"
-        )
-    return TileUrlNode(tile)
+            return self._render_url(context, url)
 
+        if tile.target:
 
-class TileUrlNode(template.Node):
-    def __init__(self, tile):
-        self.tile = template.Variable(tile)
-
-    def render(self, context):
-        tile = self.tile.resolve(context)
-        if tile.view_name:
-            try:
-                return reverse(tile.view_name)
-            except NoReverseMatch:
-                return ""
+            with context.push():
+                obj = tile.target
+                context["object"] = obj
+
+                # Template names follow typical Django naming convention, but also
+                # traverse upwards over inheritance hierarchy.
+                template_names = []
+                ct = ContentType.objects.get_for_model(obj._meta.model)
+                kls = obj._meta.model
+                while ct.model != "model":
+                    template_names.extend((
+                        "%s/inclusion_tags/%s_%s.html" % \
+                            (ct.app_label, ct.model, tile.style or "tile"),
+                        "%s/inclusion_tags/modelbase_%s.html" % \
+                            (ct.app_label, tile.style or "tile"),
+                    ))
+                    kls = kls.__bases__[0]
+                    if kls == models.Model:
+                        break
+                    ct = ContentType.objects.get_for_model(kls)
+
+                for template_name in template_names:
+                    try:
+                        return render_to_string(template_name, context.flatten())
+                    except template.TemplateDoesNotExist:
+                        pass
```

Comment:
OK, I agree: Much more solid than trying to construct a template snippet from thin air and guessing a convention. Can be customised per app/content type, for one thing. Pretty embarrassed by my code, now. I assume this is the template name resolution you followed for Jmbo?